File: 114. Flatten Binary Tree to Linked List/flattenIterative.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right

class Solution:
  def flatten(self, root: Optional[TreeNode]) -> None:
    """
    Do not return anything, modify root in-place instead.
    """
    if not root:
      return
    
    stack = [root]
    
    while stack:
      node = stack.pop()
      if node.right:
        stack.append(node.right)
      if node.left:
        stack.append(node.left)
      if stack:
        node.right = stack[-1]
      node.left = None


root = TreeNode(1)
root.left = TreeNode(2)
root.right = TreeNode(5)

root.left.left = TreeNode(3)
root.left.right = TreeNode(4)

root.right.right = TreeNode(6)
#  [1,2,3,4,5,6]
solution = Solution()
print('flatten: ', solution.flatten(root))

File: 114. Flatten Binary Tree to Linked List/flattenIterative.js
// Definition for a binary tree node.
class TreeNode {
  constructor(val = 0, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}
/**
 * @param {TreeNode} root
 * @return {void} Do not return anything, modify root in-place instead.
 */
const flatten = (root) => {
  if (!root) {
    return;
  }

  const stack = [root];

  while (stack.length) {
    const node = stack.pop();
    if (node.right) {
      stack.push(node.right);
    }
    if (node.left) {
      stack.push(node.left);
    }
    if (stack.length) {
      node.right = stack[stack.length - 1];
    }
    node.left = null;
  }
};


const root = new TreeNode(1);
root.left = new TreeNode(2);
root.right = new TreeNode(5);

root.left.left = new TreeNode(3);
root.left.right = new TreeNode(4);

root.right.right = new TreeNode(6);

console.log('flatten: ', flatten(root));
